pkg/gadgets/socket-collector/tracer/bpf/tcp4-collector.c

```diff
@@ -0,0 +1,110 @@
+// SPDX-License-Identifier: GPL-2.0 WITH Linux-syscall-note
+
+/* Copyright (c) 2021 The Inspektor Gadget authors */
+
+/*
+ * Inspired by the BPF selftests in the Linux tree:
+ * https://github.com/torvalds/linux/blob/v5.13/tools/testing/selftests/bpf/progs/bpf_iter_tcp4.c
+ */
+
+/*
+ * This BPF program uses the GPL-restricted function bpf_seq_printf().
+ */
+
+#include <vmlinux/vmlinux.h>
+
+#include <bpf/bpf_helpers.h>
+/*
+ * libbpf v0.4.0 introduced BPF_SEQ_PRINTF in bpf_tracing.h.
+ * In future versions, it will be in bpf_helpers.h.
+ */
+#include <bpf/bpf_tracing.h>
+#include <bpf/bpf_endian.h>
+#include "bpf_tracing_net.h"
+
+char _license[] SEC("license") = "GPL";
+
+static int dump_tcp_sock(struct seq_file *seq, struct tcp_sock *tp)
+{
+	const struct inet_connection_sock *icsk;
+	const struct inet_sock *inet;
+	const struct sock *sp;
+	__u16 destp, srcp;
+	__be32 dest, src;
+	int state;
+
+	icsk = &tp->inet_conn;
+	inet = &icsk->icsk_inet;
+	sp = &inet->sk;
+
+	dest = bpf_ntohl(inet->inet_daddr);
+	src = bpf_ntohl(inet->inet_rcv_saddr);
+	destp = bpf_ntohs(inet->inet_dport);
+	srcp = bpf_ntohs(inet->inet_sport);
+	state = sp->sk_state;
+
+	BPF_SEQ_PRINTF(seq, "tcp %08X %04X %08X %04X %02X\n",
+				src, srcp, dest, destp, state);
```

Comment:
I think it's ok to convert the source/destination ports to the host endianness but the IP are usually kept in the network endianness in `/proc/self/net/tcp`. Here you are converting network endianness to host endiennes in BPF and then again in Go in `func parseIPv4` so it ends up ok because `net.IP()` expects a network order. Since the logic is a bit convoluted, I suggest to add a comment before the `BPF_SEQ_PRINTF`:

```
	// Everything is printed in host endianness
	BPF_SEQ_PRINTF(seq, "tcp %08X %04X %08X %04X %02X\n",
				src, srcp, dest, destp, state);
```

ditto for other BPF_SEQ_PRINTF() calls below

Reply:
My idea was to send the ip and port in the same byte order because I didn't like how /proc/self/net/tcp mixes it. Therefore, I added the comments you suggested to make things clear. Please let me know if they are clear enough.